src/lib/transition_frontier_controller_tests/test_catchup_scheduler.ml

```diff
@@ -0,0 +1,248 @@
+          let received_rose_tree =
+            Rose_tree.T (missing_breadcrumb, dangling_rose_trees)
+          in
+          assert (
+            Rose_tree.equal received_rose_tree
+              (Rose_tree.of_list_exn upcoming_breadcrumbs)
+              ~f:(fun breadcrumb1 breadcrumb2 ->
+                External_transition.Verified.equal
+                  (With_hash.data
+                     (Transition_frontier.Breadcrumb.transition_with_hash
+                        breadcrumb1))
+                  (With_hash.data
+                     (Transition_frontier.Breadcrumb.transition_with_hash
+                        breadcrumb2)) ) ) ;
+          Strict_pipe.Writer.close catchup_breadcrumbs_writer ;
+          Strict_pipe.Writer.close catchup_job_writer )
+
+    let%test_unit "If a non-linear sequence of transitions out of order and \
+                   the missing node is received before the timeout expires, \
+                   the timeout would be canceled" =
+      let _catchup_job_reader, catchup_job_writer =
+        Strict_pipe.create Synchronous
+      in
+      let catchup_breadcrumbs_reader, catchup_breadcrumbs_writer =
+        Strict_pipe.create Synchronous
+      in
+      Thread_safe.block_on_async_exn (fun () ->
+          let open Deferred.Let_syntax in
+          let%bind frontier = setup_random_frontier () in
+          let scheduler =
+            Catchup_scheduler.create ~logger ~frontier ~time_controller
+              ~catchup_job_writer ~catchup_breadcrumbs_writer
+          in
+          let randomly_chosen_breadcrumb =
+            Transition_frontier.all_breadcrumbs frontier
+            |> List.permute |> List.hd_exn
+          in
+          let%bind upcoming_rose_tree =
+            Rose_tree.Deferred.all
+            @@ Quickcheck.random_value
+                 (gen_nonlinear_breadcrumbs ~logger ~size:5
+                    ~accounts_with_secret_keys randomly_chosen_breadcrumb)
+          in
+          let upcoming_breadcrumbs = Rose_tree.flatten upcoming_rose_tree in
+          let missing_breadcrumb = List.nth_exn upcoming_breadcrumbs 0 in
+          let missing_transition =
+            Transition_frontier.Breadcrumb.transition_with_hash
+              missing_breadcrumb
+          in
+          let dangling_breadcrumbs = List.tl_exn upcoming_breadcrumbs in
+          let dangling_transitions =
+            List.map dangling_breadcrumbs
+              ~f:Transition_frontier.Breadcrumb.transition_with_hash
+          in
+          List.iter (List.permute dangling_transitions)
+            ~f:(fun dangling_transition ->
+              Catchup_scheduler.watch scheduler ~timeout_duration
+                ~transition:dangling_transition ) ;
+          assert (not @@ Catchup_scheduler.is_empty scheduler) ;
+          Transition_frontier.add_breadcrumb_exn frontier missing_breadcrumb ;
+          Catchup_scheduler.notify scheduler ~transition:missing_transition
+          |> ignore ;
```

Comment:
It seems like this test and the previous test are very similar. Namely, you are reading from a pipe to get a rose_tree and doing some validations on the road_tree. It would be nice to create a function that dedups this.